File: 第14章-强化学习预备/04-案例/matlab/eval_baselines.m
thisDir = fileparts(mfilename('fullpath'));
root = fileparts(fileparts(fileparts(thisDir)));
addpath(fullfile(root, 'common', 'matlab'));
p = vehicle_params();
n_ep = 8;
stats = zeros(2,3); % return, steps, rmse
for kind = 1:2
    Gs = zeros(n_ep,1); Ns = Gs; Rs = Gs;
    for ep = 1:n_ep
        [G, n, rmse] = rollout_local(p, kind, 10+ep);
        Gs(ep)=G; Ns(ep)=n; Rs(ep)=rmse;
    end
    stats(kind,:) = [mean(Gs), mean(Ns), mean(Rs)];
end
fprintf('random return=%.2f steps=%.1f eyRMSE=%.3f\n', stats(1,:));
fprintf('PID    return=%.2f steps=%.1f eyRMSE=%.3f\n', stats(2,:));
figure; bar(stats(:,1)); set(gca,'XTickLabel', {'random','PID'});
ylabel('mean return'); title('Lane-keeping baselines'); grid on;

function [G, n, rmse] = rollout_local(p, kind, seed)
rng(seed);
dt = 0.05; v = 12; max_steps = 200;
ey0 = -0.5 + rand; epsi0 = -0.1 + 0.2*rand;
[Yp, psip] = path_at(0);
x = [0; Yp+ey0; psip+epsi0];
G = 0; eys = [];
for k = 1:max_steps
    [Yp, psip] = path_at(x(1));
    ey = x(2)-Yp;
    epsi = mod(x(3)-psip+pi,2*pi)-pi;
    if kind == 1
        delta = deg2rad(20)*(2*rand-1);
    else
        delta = -0.8*ey - 1.6*epsi;
    end
    delta = max(min(delta, deg2rad(35)), -deg2rad(35));
    x = bicycle_euler(x, v, delta, dt, p.L);
    r = -ey^2 - 0.5*epsi^2 - 0.05*delta^2 + 0.05;
    eys(end+1) = ey; %#ok<AGROW>
    if abs(ey) > 2.5
        r = r - 10; G = G + r; n = k; rmse = rms(eys); return;
    end
    G = G + r;
end
n = max_steps; rmse = rms(eys);
end

function [Y, psi] = path_at(X)
Y = 0.6*sin(2*pi*X/70);
dY = 0.6*(2*pi/70)*cos(2*pi*X/70);
psi = atan(dY);
end

File: 第14章-强化学习预备/04-案例/python/eval_baselines.py
from __future__ import annotations

import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np

ROOT = Path(__file__).resolve().parents[3]
sys.path.insert(0, str(ROOT / "common" / "python"))
from lane_keep_env import LaneKeepEnv  # noqa: E402


def rollout(env: LaneKeepEnv, policy, seed: int):
    obs = env.reset(seed=seed)
    G = 0.0
    eys = []
    for _ in range(env.max_steps):
        act = policy(obs)
        obs, r, done, info = env.step(act)
        G += r
        eys.append(info["ey"])
        if done:
            break
    return G, len(eys), float(np.sqrt(np.mean(np.square(eys))))


def pid_policy(obs):
    ey, epsi, _ = obs
    return -0.8 * ey - 1.6 * epsi


def random_policy(obs, rng):
    return rng.uniform(-np.deg2rad(20), np.deg2rad(20))


def main() -> None:
    env = LaneKeepEnv()
    rng = np.random.default_rng(0)
    n_ep = 8
    rows = []
    for name, pol in [
        ("random", lambda o: random_policy(o, rng)),
        ("PID", pid_policy),
    ]:
        Gs, steps, rmses = [], [], []
        for ep in range(n_ep):
            G, n, rmse = rollout(env, pol, seed=10 + ep)
            Gs.append(G)
            steps.append(n)
            rmses.append(rmse)
        print(f"{name:6s}  return={np.mean(Gs):7.2f}  steps={np.mean(steps):6.1f}  eyRMSE={np.mean(rmses):.3f}")
        rows.append((name, Gs, rmses))

    fig, ax = plt.subplots(figsize=(6, 4))
    ax.bar([0, 1], [np.mean(rows[0][1]), np.mean(rows[1][1])], color=["0.6", "0.2"])
    ax.set_xticks([0, 1], ["random", "PID"])
    ax.set_ylabel("mean return")
    ax.set_title("Lane-keeping baselines (higher is better)")
    ax.grid(True, axis="y")
    fig.tight_layout()
    out = Path(__file__).with_name("eval_baselines.png")
    fig.savefig(out, dpi=120)
    print(f"saved {out}")
    plt.show()


if __name__ == "__main__":
    main()
